### PR TITLE
Рождественский Павел, 2536

### PR DESCRIPTION
**Pull request by PaulChristmas**

### CODE REVIEW

**Comment by thevery** [comment]
- no toast on adding non-existing city
- city list on start is so-so for usability
- no error on failed update
- last update time format is ugly

**Comment by PaulChristmas** [comment]
"ugly" is too subjective thing. and what means "so-so for usability" ?

File: app/src/main/java/ru/ifmo/md/lesson8/CityDetailFragment.java [hunk 1]
```diff
@@ -0,0 +1,238 @@
...
+        if (mItemCursor == null) {
+            return;
+        }
+        long start = System.currentTimeMillis();
+        final int currentWeatherIconId = getImageById(mItemCursor.getString(mItemCursor.getColumnIndex(WeatherContract.City.WEATHER_ICON_ID)));
+        final String weatherDesc = mItemCursor.getString(mItemCursor.getColumnIndex(WeatherContract.City.WEATHER_DESCRIPTION));
+        final String cityName = mItemCursor.getString(mItemCursor.getColumnIndex(WeatherContract.City.CITY_NAME));
+        final double temp = mItemCursor.getDouble(mItemCursor.getColumnIndex(WeatherContract.City.CITY_TEMPERATURE));
+        final String temperature = String.format("%.1f°", temp);
+        final String lastUpdate = "Last update: " + mItemCursor.getString(mItemCursor.getColumnIndex(WeatherContract.City.CITY_LAST_UPDATE)).replace("T", " ");
```
Comment (thevery) [inline app/src/main/java/ru/ifmo/md/lesson8/CityDetailFragment.java:142]: `strings.xml`!

File: app/src/main/java/ru/ifmo/md/lesson8/CityDetailFragment.java [hunk 2]
```diff
@@ -0,0 +1,238 @@
...
+        }
+        long tm = System.currentTimeMillis() - start;
+        Log.d("TAG", "Updating UI took " + tm + " ms");
+    }
+    private String toTitleCase(String s) {
+        return String.valueOf(Character.toUpperCase(s.charAt(0))) + s.subSequence(1, s.length()).toString().toLowerCase();
+    }
+    private int getImageById(String icon) {
+        switch (icon) {
+            case "01d": return R.drawable.w01d;
```
Comment (thevery) [inline app/src/main/java/ru/ifmo/md/lesson8/CityDetailFragment.java:217]: `getIdentifier()`

File: app/src/main/java/ru/ifmo/md/lesson8/WeatherLoaderService.java [hunk 3]
```diff
@@ -0,0 +1,394 @@
...
+            alarmManager.setRepeating(AlarmManager.RTC, System.currentTimeMillis(),
+                    readInterval(context), pi);
+        } else {
+            alarmManager.cancel(pi);
+            pi.cancel();
+        }
+    }
+    private static long readInterval(Context context) {
+        return PreferenceManager.getDefaultSharedPreferences(context)
+                .getLong("interval", INTERVAL_HOUR);
```
Comment (thevery) [inline app/src/main/java/ru/ifmo/md/lesson8/WeatherLoaderService.java:91]: const!

File: app/src/main/java/ru/ifmo/md/lesson8/WeatherLoaderService.java [hunk 4]
```diff
@@ -0,0 +1,394 @@
...
+                .commit();
+    }
+    public static boolean isServiceAlarmOn(Context context) {
+        Intent i = new Intent(context, WeatherLoaderService.class);
+        PendingIntent pi = PendingIntent.getService(context, 0, i, PendingIntent.FLAG_NO_CREATE);
+        return pi != null;
+    }
+    public static int getIntervalIndex(Context context) {
+        final long interval = readInterval(context);
+        if (interval == INTERVAL_HOUR) return 1;
```
Comment (thevery) [inline app/src/main/java/ru/ifmo/md/lesson8/WeatherLoaderService.java:116]: no 1-line ifs, please!

File: app/src/main/java/ru/ifmo/md/lesson8/WeatherLoaderService.java [hunk 5]
```diff
@@ -0,0 +1,394 @@
...
+            }
+        }
+    }
+    private void handleActionUpdateAll() {
+        Cursor cursor = getContentResolver().query(
+                WeatherContract.City.CONTENT_URI,
+                WeatherContract.City.ID_COLUMNS,
+                null, null, null);
+        cursor.moveToFirst();
+        while (!cursor.isBeforeFirst() && !cursor.isAfterLast()) {
```
Comment (thevery) [inline app/src/main/java/ru/ifmo/md/lesson8/WeatherLoaderService.java:160]: `while (cursor.moveToNext())`

File: app/src/main/java/ru/ifmo/md/lesson8/WeatherLoaderService.java [hunk 6]
```diff
@@ -0,0 +1,394 @@
...
+            builder.append(day.getMaxTemperature()).append("$");
+        }
+        cv.put(WeatherContract.City.WEATHER_FORECAST, builder.toString());
+        return cv;
+    }
+    private InputStreamReader getReader(String url) throws IOException {
+        URL targetUrl = new URL(url);
+        HttpURLConnection connection = (HttpURLConnection) targetUrl.openConnection();
+        InputStream is = connection.getInputStream();
+        return new InputStreamReader(is);
```
Comment (thevery) [inline app/src/main/java/ru/ifmo/md/lesson8/WeatherLoaderService.java:280]: close?

**Comment by thevery** [comment]
> "ugly" is too subjective thing.

"ugly" because of too much details instead of "5 min. ago", "11:27", "yesterday", "3 days ago", etc

**Comment by thevery** [comment]
> and what means "so-so for usability" ?

Weather apps used to check weather, not the list of added cities, so it makes sense to open first/last used/current/etc city directly.

File: app/src/main/java/ru/ifmo/md/lesson8/WeatherLoaderService.java [hunk 7]
```diff
@@ -0,0 +1,394 @@
...
+        cv.put(WeatherContract.City.CITY_HUMIDITY, weather.getHumidity());
+        cv.put(WeatherContract.City.CITY_WIND_SPEED, weather.getWindSpeed());
+        cv.put(WeatherContract.City.WEATHER_DESCRIPTION, weather.getWeatherDescription());
+        cv.put(WeatherContract.City.WEATHER_ICON_ID, weather.getIcon());
+        StringBuilder builder = new StringBuilder();
+        for (WeatherDay day : weather.getForecast()) {
+            builder.append(day.getDate()).append("$");
+            builder.append(day.getIcon()).append("$");
+            builder.append(day.getMinTemperature()).append("$");
+            builder.append(day.getMaxTemperature()).append("$");
```
Comment (thevery) [inline app/src/main/java/ru/ifmo/md/lesson8/WeatherLoaderService.java:271]: ugly hack

**Comment by thevery** [comment]
- add city EditText is not single line

ok
24-7=17 => 9

**Pull request closed without merging**